src/btree/bt_page.c

```diff
@@ -418,20 +454,178 @@ __wt_page_inmem(WT_SESSION_IMPL *session, WT_REF *ref, const void *image, uint32
     return (ret);
 }
 
+/*
+ * __wt_col_fix_read_auxheader --
+ *     Read the auxiliary header following the bitmap data, if any. This code is used by verify and
+ *     needs to be accordingly careful. It is also used by mainline reads so it must also not crash
+ *     or print.
+ */
+int
+__wt_col_fix_read_auxheader(
+  WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk, WT_COL_FIX_AUXILIARY_HEADER *auxhdr)
+{
+    WT_BTREE *btree;
+    uint32_t bitmapsize, auxheaderoffset;
+    const uint8_t *raw, *end;
+
+    btree = S2BT(session);
+
+    /*
+     * Figure where the auxiliary header is. It is always immediately after the bitmap data,
+     * regardless of whether the page is full.
+     */
+    bitmapsize = __bitstr_size(dsk->u.entries * btree->bitcnt);
+    auxheaderoffset = WT_PAGE_HEADER_BYTE_SIZE(btree) + bitmapsize;
+
+    /*
+     * If the auxiliary header is past the in-memory page size, there's no auxiliary data. If
+     * there's at least one byte past the bitmap data, check whether it's zero. If that's zero,
+     * there's no auxiliary data. (We are guaranteed that any allocation slop that we might be
+     * looking at is all zeros.) Set everything to zero and return.
+     */
+    if (auxheaderoffset >= dsk->mem_size || *(raw = (uint8_t *)dsk + auxheaderoffset) == 0) {
+        auxhdr->version = WT_COL_FIX_VERSION_NIL;
+        auxhdr->entries = 0;
+        auxhdr->offset = 0;
+        return (0);
+    }
+
+    /* Remember the end of the page for easy computation of maximum lengths. */
+    end = (uint8_t *)dsk + dsk->mem_size;
+
+    /*
+     * The on-disk header is a 1-byte version, a uint32 with the number of entries, and a uint32
+     * that says how many waste bytes there are between the header and the entries.
+     */
+    auxhdr->version = raw[0];
+    raw++;
+    {
+        uint32_t entries, waste;
```

Comment:
That's why I put it in its own block. Since I guess that's also illegal, I'll uglify it. :-)  (Fixed) (This code then got rearranged more regarding bounds checks and it's no longer particularly ugly.)